src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -637,6 +609,7 @@ public void start() throws ConfigurationException, InterruptedException, IOExcep
   @PreDestroy
   public void stop() throws InterruptedException {
     System.err.println("*** shutting down gRPC server since JVM is shutting down");
+    prepareWorkerForGracefulShutdown();
```

Comment:
my understanding is, this will work when the worker process received interruption. Will the new mechanism work in a scale in cycle of a **non k8s** environment? i was expecting autoscaling from aws would just terminated the worker instance instead of waiting for the java process to exit properly.

Reply:
It should work with AWS's lifecycle hook. There will most likely be additional configuration needed on the EC2 side to catch the hook and issue a sigterm to stop the running buildfarm worker process. We did it in the past by writing an empty .paused file but that was a bit hacky. This is a better solution that does not require writing files but rather just stop the process and let it handle the rest.

Reply:
I see. yeah sounds like we want to add comments and docs about the fact that this will rely on the lifecycle hook check feature of aws. This will make things easier for users who want to use the feature.

Reply:
Just to clarify, it always did rely on the lifecycle hook, which is how we used it in the past. Essentially, the lifecycle hook would send a signal to the EC2 instance, which then would write the .paused file which de-registed the worker. Then lifecycle hook would send the termination signal after the hook period time elapsed. There is no change to that use case here.